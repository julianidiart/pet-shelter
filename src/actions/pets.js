import uuid from "uuid";
import database from "../firebase";

// ADD_PET
export const addPet = pet => ({
  type: "ADD_PET",
  pet
});

export const startAddPet = (petData = {}) => {
  return dispatch => {
    const { name = "" } = petData;
    const pet = { name };
    database
      .ref("pet")
      .push(pet)
      .then(ref => {
        dispatch(addPet({ id: ref.key, ...pet }));
      });
  };
};

// REMOVE_PET
export const removePet = id => ({
  type: "REMOVE_PET",
  id
});

// EDIT_PET
export const editPet = (id, pet) => ({
  type: "EDIT_PET",
  id,
  pet
});
