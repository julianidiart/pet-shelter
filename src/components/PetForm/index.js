import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

export default class PetForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: props.pet ? props.pet.name : "",
      birthdate: props.pet ? moment(props.pet.birthdate) : moment(),
      chip: props.pet ? props.pet.chip : "",
      place: props.pet ? props.pet.place : "",
      sex: props.pet ? props.pet.sex : "",
      breed: props.pet ? props.pet.breed : "",
      calendarFocused: false,
      error: ""
    };
  }

  onInputChange = e => {
    const target = e.target;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    this.setState(() => ({ [name]: value }));
  };

  onDateChange = birthdate => {
    if (birthdate) {
      this.setState(() => ({ birthdate }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (this.state.name) {
      const error = "";
      this.setState(() => ({ error }));
      this.props.onSubmit({
        name: this.state.name,
        birthdate: this.state.birthdate.valueOf(),
        chip: this.state.chip,
        place: this.state.place,
        sex: this.state.sex,
        breed: this.state.breed
      });
    } else {
      const error = "The pet should have a name!";
      this.setState(() => ({ error }));
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            autoFocus
            onChange={this.onInputChange}
            placeholder="Name"
            type="text"
            value={this.state.name}
            name="name"
          />
          <SingleDatePicker
            date={this.state.birthdate}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            displayFormat={() => "DD/MM/YYYY"}
          />
          <input
            autoFocus
            onChange={this.onInputChange}
            placeholder="Sex"
            type="text"
            value={this.state.sex}
            name="sex"
          />
          <input
            autoFocus
            onChange={this.onInputChange}
            placeholder="Breed"
            type="text"
            value={this.state.breed}
            name="breed"
          />
          <input
            autoFocus
            onChange={this.onInputChange}
            placeholder="Chip"
            type="text"
            value={this.state.chip}
            name="chip"
          />
          <input
            autoFocus
            onChange={this.onInputChange}
            placeholder="Place"
            type="text"
            value={this.state.place}
            name="place"
          />
          <button>Save Pet</button>
        </form>
      </div>
    );
  }
}
