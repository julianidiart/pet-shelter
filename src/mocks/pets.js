export default [
  { id: "1", name: "Marla" },
  { id: "2", name: "Bimba" },
  { id: "3", name: "Aimee" },
  { id: "4", name: "Maia" },
  { id: "5", name: "Emma" },
  { id: "6", name: "Tommy" },
  { id: "7", name: "Bello Bruno" },
  { id: "8", name: "Bubu" },
  { id: "9", name: "Blaky" }
];
