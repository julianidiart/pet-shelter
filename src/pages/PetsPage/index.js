import React from "react";
import PetsSummary from "../../components/PetsSummary";
import PetsListFilters from "../../components/PetsList/PetsListFilters";
import PetsList from "../../components/PetsList";

export default props => (
  <div>
    <PetsSummary />
    <PetsListFilters />
    <PetsList />
  </div>
);
