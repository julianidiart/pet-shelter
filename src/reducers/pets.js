const petsReducerDefaultState = [];

export default (state = petsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_PET":
      return [...state, action.pet];
    case "REMOVE_PET":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_PET":
      return state.map(pet => {
        if (pet.id === action.id) {
          return {
            ...pet,
            ...action.pet
          };
        } else {
          return pet;
        }
      });
    case "SET_PETS":
      return action.pets;
    default:
      return state;
  }
};
