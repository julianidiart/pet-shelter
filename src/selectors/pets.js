import moment from "moment";

export default (pets, { text, sortBy, startDate, endDate, filter }) => {
  return pets
    .filter(pet => {
      const birthdateMoment = moment(pet.birthdate);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(birthdateMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(birthdateMoment, "day")
        : true;
      const textMatch =
        pet.name.toLowerCase().includes(text.toLowerCase()) ||
        pet.place.toLowerCase().includes(text.toLowerCase()) ||
        pet.breed.toLowerCase().includes(text.toLowerCase()) ||
        (pet.comments &&
          pet.comments.toLowerCase().includes(text.toLowerCase())) ||
        (pet.color
          ? pet.color.toLowerCase().includes(text.toLowerCase())
          : false) ||
        pet.chip
          .toString()
          .toLowerCase()
          .includes(text.toLowerCase());
      let filterMatch = true;
      const adoptedOrPassedAway = pet.adopted || pet.passedAway;
      switch (filter) {
        case "sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === true;
          break;
        case "non-sterilized":
          filterMatch = !adoptedOrPassedAway && pet.sterilized === false;
          break;
        case "chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip !== "";
          break;
        case "non-chipped":
          filterMatch = !adoptedOrPassedAway && pet.chip === "";
          break;
        case "adopted":
          filterMatch = pet.adopted === true;
          break;
        case "passed-away":
          filterMatch = pet.passedAway === true;
          break;
        default:
          filterMatch = !adoptedOrPassedAway;
      }

      return startDateMatch && endDateMatch && textMatch && filterMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.birthdate < b.birthdate ? 1 : -1;
      }
      return a.name < b.name ? 1 : -1;
    });
};
