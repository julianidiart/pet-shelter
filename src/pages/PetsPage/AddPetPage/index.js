import React, { Component } from "react";
import { connect } from "react-redux";
import PetForm from "../../../components/PetForm";
import { startAddPet } from "../../../actions/pets";

export class AddPetPage extends Component {
  onSubmit = pet => {
    console.log(pet);
    this.props.startAddPet(pet);
    this.props.history.push("/pets");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Pet</h1>
          </div>
        </div>
        <div className="content-container">
          <PetForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPet: pet => dispatch(startAddPet(pet))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPetPage);
