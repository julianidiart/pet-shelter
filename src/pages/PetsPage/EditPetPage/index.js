import React, { Component } from "react";
import { connect } from "react-redux";
import PetForm from "../../../components/PetForm";
import { startEditPet, startRemovePet } from "../../../actions/pets";
import MultiLanguageText from "../../../components/MultiLanguageText";

export class EditPetPage extends Component {
  onSubmit = pet => {
    this.props.startEditPet(this.props.pet.id, pet);
    this.props.history.push("/pets");
  };
  onClick = () => {
    this.props.startRemovePet({ id: this.props.pet.id });
    this.props.history.push("/pets");
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">
              <MultiLanguageText
                en="Add Pet"
                it="Aggiungi Animale"
                es="Agregar Mascota"
              />
            </h1>
          </div>
        </div>
        <div className="content-container">
          <PetForm pet={this.props.pet} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onClick}>
            <MultiLanguageText
              en="Remove Pet"
              it="Rimuovere Animale"
              es="Eliminar Mascota"
            />
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    pet: state.pets.find(pet => pet.id === props.match.params.id)
  };
};

const mapDispatchToProps = dispatch => ({
  startEditPet: (id, pet) => dispatch(startEditPet(id, pet)),
  startRemovePet: id => dispatch(startRemovePet(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPetPage);
