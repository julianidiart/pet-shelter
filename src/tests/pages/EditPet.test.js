import React from "react";
import { shallow } from "enzyme";
import pets from "../fixtures/pets";
import { EditPetPage } from "../../pages/PetsPage/EditPetPage";
import PetForm from "../../components/PetForm";

let editPet, removePet, history, wrapper;

beforeEach(() => {
  editPet = jest.fn();
  removePet = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditPetPage
      editPet={editPet}
      removePet={removePet}
      history={history}
      pet={pets[0]}
    />
  );
});

test("should render EditPetPage", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editPet", () => {
  wrapper.find(PetForm).prop("onSubmit")(pets[0]);
  expect(history.push).toHaveBeenLastCalledWith("/pets");
  expect(editPet).toHaveBeenLastCalledWith(pets[0].id, pets[0]);
});

test("should handle removePet", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/pets");
  expect(removePet).toHaveBeenLastCalledWith(pets[0].id);
});
