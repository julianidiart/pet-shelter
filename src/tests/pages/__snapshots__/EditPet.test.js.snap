// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditPetPage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Pet
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <PetForm
      onSubmit={[Function]}
      pet={
        Object {
          "avatarURL": "",
          "birthdate": -63158400000,
          "breed": "Incrocio",
          "chip": "123ABC123",
          "color": "Black and white",
          "id": "1",
          "name": "Marla",
          "place": "Barchetta",
          "sex": "F",
          "size": "M",
          "sterilized": true,
        }
      }
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Pet
    </button>
  </div>
</div>
`;
